[FIX] html_builder, web_editor, website: background filters corner radius fixed

dabo-odoo opened the pull request:
This commit aims to replace https://github.com/odoo/odoo/commit/4d9f6f664e84216ed505c4b860f8ac9f210ca4fb by using CSS to calculate the inner radius of the background layer instead of JS

task-3358501

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/216069.png)](https://mergebot.odoo.com/odoo/odoo/pull/216069)

dabo-odoo commented on addons/html_builder/static/src/core/utils.js:
```diff
@@ -659,7 +659,9 @@ export function useInputBuilderComponent({
     function commit(userInputValue) {
         if (defaultValue !== undefined) {
-            userInputValue ||= formatRawValue(defaultValue);
+            if (!userInputValue || (typeof userInputValue === "string" && !userInputValue.trim())) {
+                userInputValue = formatRawValue(defaultValue);
+            }
```
Moved to another commit with explanation

dabo-odoo commented on addons/html_builder/static/src/core/border_radius_style_plugin.js:
```diff
@@ -1,40 +1,21 @@
-                this.calculateAndSetRadius(gridEl);
-            });
-        },
+        apply_custom_css_style: withSequence(20, this.applyRadiusStyle.bind(this)),
```
It gets called in core_builder_action_plugin every time a StyleAction is applied. The plugin itself is not needed, I figured that putting the applyRadiusStyle resource in a plugin called border_radius_whatever would be easier to find, but I can move it without problems if needed.

qsm-odoo commented:
@dabo-odoo Forgot to say: also, your commit message do not mention the right apps and it should mention merged commits, not PR links.

dabo-odoo commented:
@qsm-odoo Made it so that border-width/radius styles get converted to --box-border variables so that it works on existing snippets as well.
Also I found out that the issue that the first commit fixes is introduced by the other commit in this PR; for now I just reordered them, I don't know if I should keep two commits or squash them into one.

qsm-odoo commented on addons/html_builder/static/src/plugins/border_configurator_option.js:
```diff
@@ -20,6 +21,38 @@ export class BorderConfigurator extends BaseOptionComponent {
             hasBorder: this.hasBorder(editingElement),
         }));
+
+        // Convert border-width/radius styles into --box-border-width/radius
+        // variables if border uses Bootstrap classes
+        if (this.props.withBSClass) {
+            const el = this.env.getEditingElement();
+            if (el.style["border-width"]) {
+                setStyle(el, "--box-border-width", el.style["border-width"], {
```
Maybe that changed in the new editor but I doubt it is a good idea for an option `setup` to change the website DOM. Even functionally, this really should be reserved to users **using options**.

What I would expect is the switch to the variable use to be done only when the user enter a new value for the option. So just at the place where you do "set --box-border-width to Xpx", change into "set --box-border-width to Xpx and remove the inline style border-width".

qsm-odoo commented on addons/web/static/lib/bootstrap/scss/_utilities.scss:
```diff
@@ -709,7 +709,7 @@ $utilities: map-merge(
     // scss-docs-start utils-border-radius
     "rounded": (
-      property: border-radius,
+      property: --box-border-radius,
```
You cannot change the lib file directly. You can change this map by modifying it at the right place: I see we have a `utilities_custom` stuff, it should probably be a specific one for the frontend, in web_editor (or in the future html_builder I guess).

qsm-odoo commented on addons/web_editor/static/src/scss/web_editor.common.scss:
```diff
@@ -759,6 +759,40 @@ pre {
+        var(--box-border-left-width, var(--box-border-width, var(--border-width))) !important;
+}
+
+.rounded {
+    border-radius:
+        var(--box-border-top-left-radius, var(--box-border-radius, var(--border-radius)))
+        var(--box-border-top-right-radius, var(--box-border-radius, var(--border-radius)))
+        var(--box-border-bottom-right-radius, var(--box-border-radius, var(--border-radius)))
+        var(--box-border-bottom-left-radius, var(--box-border-radius, var(--border-radius))) !important;
+}
+
+.rounded-1, .rounded-2, .rounded-3, .rounded-4, .rounded-5, .rounded-circle, .rounded-pill,
+.rounded-top-1, .rounded-top-2, .rounded-top-3, .rounded-top-4, .rounded-top-5, .rounded-top-circle, .rounded-top-pill,
+.rounded-end-1, .rounded-end-2, .rounded-end-3, .rounded-end-4, .rounded-end-5, .rounded-end-circle, .rounded-end-pill,
+.rounded-bottom-1, .rounded-bottom-2, .rounded-bottom-3, .rounded-bottom-4, .rounded-bottom-5, .rounded-bottom-circle, .rounded-bottom-pill,
+.rounded-start-1, .rounded-start-2, .rounded-start-3, .rounded-start-4, .rounded-start-5, .rounded-start-circle, .rounded-start-pill {
+    @extend .rounded;
 }
```
It's basically never a good idea to use `@extend` on a normal selector. In this case: why do you want those classe to behave like `.rounded` in ALL customs selector there could be? If a customer (or us in an app) do something very simple like this:
```scss
.my-class.rounded {
    rules;

    .child1 {
          rules2;
    }
    .child2 {
          rules3;
    }
}
```
This will increase all those rules and child subrules to use not only rounded but all your classes that extend it. That's not what you want here.

If you want to create a unique selector for a common set of rules, you can use `@extend` but with a [*placeholder selector*](https://sass-lang.com/documentation/at-rules/extend/#placeholder-selectors). Like this:
```scss
%this-is-a-placeholder-selector {
    rules;
}

selectors {
    @extend %this-is-a-placeholder-selector;
}
```
=> You ensure to inherit only what you need with this `extend` 


But in this case, it's even more simple: why do you use `extend` at all? Can't you just put the `rounded` class in your big list and set up the CSS property directly?

But in the end, don't make a ugly big hardcoded selector like that: generate it by looping through the right `$utilities` map keys. And in that case, you will be able to use the placeholder extend trick:
```scss
%box-border-radius-rules {
    the-rules;
}

$utilities-radius-keys: ('rounded', 'rounded-top', ...)
// then loop through $utilites and do `@extend %box-border-radius-rules`
```

Also please add comment next to this to explain the concept and that bootstrap is overridden to use the CSS variable, etc.

dabo-odoo commented:
@qsm-odoo Implemented the changes but I had to squash the userInputValue fix or the runbot would not pass

dabo-odoo commented:
@qsm-odoo
- moved the commit to be the first
- changed explanation to be independent of second commit
- changed steps to reproduce to be similar to your example
- added a test to check that inserting a whitespace reverts back to the previous value

Let me know if I need to change anything else

qsm-odoo commented on addons/website/static/tests/builder/options/border_configurator_option.test.js:
```diff
@@ -32,9 +32,24 @@ test("empty border input is treated as 0", async () => {
     await contains(":iframe section").click();
     expect.verifySteps(["hasBorder"]);
-    await contains(".options-container [data-label=Border] input").edit(" "); // .clear() doesn't trigger a rerender.
+    await contains(".options-container [data-label=Border] input").edit(""); // .clear() doesn't trigger a rerender.
     expect.verifySteps(["hasBorder"]);
```
@bso-odoo That still is broken :shrug: Anyways, I'll just add a comment and be done with it. Actually, by fixing the last commit, the test actually fails there too... it's consistent at least :shrug:

qsm-odoo reviewed (changes requested):
Dropping comments, this still does not work. I am push-forcing with my adaptations, I'll fix with a clearer head tomorrow :grimacing:

robodoo commented:
Merge method set to rebase and fast-forward.
